tests/api/api/test_artifacts.py

```diff
@@ -264,6 +264,60 @@ def test_list_artifacts(db: Session, client: TestClient) -> None:
     )
 
 
+def test_list_artifacts_with_limits(
+    db: Session, unversioned_client: TestClient
+) -> None:
+    _create_project(unversioned_client, prefix="v1")
+
+    def ensure_endpoint_limit(limit: int):
+        for route in unversioned_client.app.routes:
+            if route.path.endswith(LIST_API_ARTIFACTS_V2_PATH):
+                for qp in route.dependant.query_params:
+                    if qp.name == "limit":
+                        qp.default = limit
+                        break
+
+    default_endpoint_limit = 2
+    ensure_endpoint_limit(default_endpoint_limit)
+
+    for i in range(default_endpoint_limit + 1):
+        data = {
+            "kind": "artifact",
+            "metadata": {
+                "description": "",
+                "labels": {},
+                "key": KEY,
+                "project": PROJECT,
+                "tree": str(uuid.uuid4()),
+            },
+            "spec": {
+                "db_key": "some-key",
+                "producer": {"kind": "api"},
+                "target_path": "memory://aaa/aaa",
+            },
+            "status": {},
+        }
+        resp = unversioned_client.post(
+            STORE_API_ARTIFACTS_V2_PATH.format(project=PROJECT),
+            json=data,
+        )
+        assert resp.status_code == HTTPStatus.CREATED.value
+
+    artifact_path = LIST_API_ARTIFACTS_V2_PATH.format(project=PROJECT)
+    resp = unversioned_client.get(f"{artifact_path}?limit={default_endpoint_limit-1}")
+    assert resp.status_code == HTTPStatus.OK.value
+    artifacts = resp.json()["artifacts"]
+    assert len(artifacts) == default_endpoint_limit - 1
+
+    # Get all artifacts
+    resp = unversioned_client.get(artifact_path)
+    assert resp.status_code == HTTPStatus.OK.value
+    artifacts = resp.json()["artifacts"]
+    assert len(artifacts) == default_endpoint_limit
+
+    ensure_endpoint_limit(1000)
```

Comment:
this won't run if an `assert` fails.
Use a fixture to achieve `defer`-like behavior